src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,867 @@
+                operator: operator,
+                strategy: allocations[i].strategy,
+                numToComplete: type(uint16).max
+            });
+
+            // 2. Check current totalMagnitude matches expected value. This is to check for slashing race conditions
+            // where an operator gets slashed from an operatorSet and as a result all the configured allocations have larger
+            // proprtional magnitudes relative to eachother. This check prevents any surprising behavior.
+            uint64 currentTotalMagnitude = _getLatestTotalMagnitude(operator, allocations[i].strategy);
+            require(currentTotalMagnitude == allocations[i].expectedTotalMagnitude, InvalidExpectedTotalMagnitude());
+
+            // 3. set allocations for the strategy after updating freeMagnitude
+            _modifyAllocations({
+                operator: operator,
+                allocation: allocations[i],
+                allocationEffectTimestamp: effectTimestamp,
+                deallocationCompletableTimestamp: completableTimestamp
+            });
+        }
+    }
+
+    /**
+     * @notice Called by an AVS to slash an operator for given operatorSetId, list of strategies, and bipsToSlash.
+     * For each given (operator, operatorSetId, strategy) tuple, bipsToSlash
+     * bips of the operatorSet's slashable stake allocation will be slashed
+     *
+     * @param operator the address to slash
+     * @param operatorSetId the ID of the operatorSet the operator is being slashed on behalf of
+     * @param strategies the set of strategies to slash
+     * @param bipsToSlash the number of bips to slash, this will be proportional to the
+     * operator's slashable stake allocation for the operatorSet
+     */
+    function slashOperator(
+        address operator,
+        uint32 operatorSetId,
+        IStrategy[] calldata strategies,
+        uint16 bipsToSlash
+    ) external {
+        OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: operatorSetId});
+        bytes32 operatorSetKey = _encodeOperatorSet(operatorSet);
+        require(isOperatorSlashable(operator, operatorSet), InvalidOperator());
+
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            // 1. calculate slashed magnitude from current allocation
+            // update current and all following queued magnitude updates for (operator, strategy, operatorSetId) tuple
+            uint64 slashedMagnitude;
+            {
+                uint64 currentMagnitude = uint64(
+                    _magnitudeUpdate[operator][strategies[i]][operatorSetKey].upperLookupRecent(uint32(block.timestamp))
+                );
+                // slash nonzero current magnitude
+                if (currentMagnitude != 0) {
+                    /// TODO: add wrapping library for rounding up for slashing accounting
+                    slashedMagnitude = uint64(uint256(bipsToSlash) * uint256(currentMagnitude) / BIPS_FACTOR);
```

Comment:
pretty sure this will round up properly if you just add `9999` (i.e. `BIPS_FACTOR-1` prior to dividing)
so make it:
```
                    slashedMagnitude = uint64((uint256(bipsToSlash) * uint256(currentMagnitude) + BIPS_FACTOR - 1) / BIPS_FACTOR);
```

Reply:
Will make a pr to check/improve rounding where needed.